Work around error in <Block_private.h> with open source clang

ddkilzer opened the pull request:
#### d34db4fc6f99c2c5d5234bcdc45b3c120f91e808
<pre>
Work around error in &lt;Block_private.h&gt; with open source clang
<a href="https://bugs.webkit.org/show_bug.cgi?id=267197">https://bugs.webkit.org/show_bug.cgi?id=267197</a>
&lt;<a href="https://rdar.apple.com/120600666">rdar://120600666</a>&gt;

Reviewed by Alexey Proskuryakov and Darin Adler.

* Source/WTF/wtf/Compiler.h:
- Define COMPILER(APPLE_CLANG) when building with Apple&apos;s clang.
* Source/WebKit/Platform/spi/Cocoa/BlockSPI.h:
- Do not include &lt;Block_private.h&gt; when compiling with open source
  clang to work around the error.

Canonical link: <a href="https://commits.webkit.org/272793@main">https://commits.webkit.org/272793@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb39f76dbfa899dd2fbbd4a84c9eff3938b68876

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32925 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11683 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35549 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29746 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8859 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29189 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33381 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9847 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29407 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8581 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8729 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36881 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/28174 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29904 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29772 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34862 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/32933 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6810 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32716 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10550 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/39395 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9463 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8300 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4255 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9484 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKit/Platform/spi/Cocoa/BlockSPI.h:
```diff
@@ -23,7 +23,8 @@
+// FIXME: COMPILER(APPLE_CLANG) is needed until <rdar://120571425> is fixed.
+#if USE(APPLE_INTERNAL_SDK) && COMPILER(APPLE_CLANG)
 
 #import <Block_private.h>
```
If we can compile properly without this include, perhaps it can be just removed?

ddkilzer replied:
> If we can compile properly without this include, perhaps it can be just removed?

That's basically true for any internal/private header.  :)

We've always preferred the private header over redeclaring a function when compiling with the internal SDK.

The `<Block_private.h>` header is declaring this function (which is re-declared for the public SDK):

```
BLOCK_EXPORT const char* _Block_signature(void*);
```

(Not sure why GitHub won't let you expand the rest of the file, though.)

aproskuryakov replied:
Makes sense, I didn't see that it was redeclared below.

darinadler commented on Source/WebKit/Platform/spi/Cocoa/BlockSPI.h:
```diff
@@ -23,7 +23,8 @@
  */
 
-#if USE(APPLE_INTERNAL_SDK)
+// FIXME: COMPILER(APPLE_CLANG) is needed until <rdar://120571425> is fixed.
```
Seems like this could just use `defined(__apple_build_version__)`; I don’t see much value in the indirection.

ddkilzer replied:
I think this:

```
#if USE(APPLE_INTERNAL_SDK) && COMPILER(APPLE_CLANG)
```

Reads better than this (need to also add the `defined(__clang__)` check for context):

```
#if USE(APPLE_INTERNAL_SDK) && defined(__clang__) && defined(__apple_build_version__)
```

On the other hand, defining `WTF_COMPILER_APPLE_CLANG` in `Compiler.h` is self-documenting because of where it happens:

```
#if defined(__clang__)
[...]
#if defined(__apple_build_version__)
#define WTF_COMPILER_APPLE_CLANG 1
#endif
[...]
#endif
```

darinadler replied:
OK, no worries

Commit:
Work around error in <Block_private.h> with open source clang
https://bugs.webkit.org/show_bug.cgi?id=267197
<rdar://120600666>

Reviewed by Alexey Proskuryakov and Darin Adler.

* Source/WTF/wtf/Compiler.h:
- Define COMPILER(APPLE_CLANG) when building with Apple's clang.
* Source/WebKit/Platform/spi/Cocoa/BlockSPI.h:
- Do not include <Block_private.h> when compiling with open source
  clang to work around the error.

Canonical link: https://commits.webkit.org/272793@main
